In [ ]:
import requests

In [ ]:
import bs4

In [ ]:
from bs4 import BeautifulSoup

In [ ]:
'''
urls = [
    'https://en.wikipedia.org/wiki/Web_scraping',
    'https://en.wikipedia.org/wiki/Machine_learning',
    'https://en.wikipedia.org/wiki/Artificial_intelligence',
    'https://en.wikipedia.org/wiki/Gradient_descent'
]
with open('Wikipedia_scarp.txt','w') as f:
  for url in urls:
  # Send a GET request to the URL and retrieve the HTML content
    response = requests.get(url)
    html_content = response.content

    # Parse the HTML content using Beautiful Soup
    soup = BeautifulSoup(html_content, 'html.parser')

    # Find the main content of the page (typically inside a div or section tag)
    content_div = soup.find('div', {'id': 'mw-content-text'})

    # Extract the text from the content div
    text = content_div.get_text()

    # Print the scraped text
    f.write(text)


print('Scraped text has been written to file')
'''

"\nurls = [\n    'https://en.wikipedia.org/wiki/Web_scraping',\n    'https://en.wikipedia.org/wiki/Machine_learning',\n    'https://en.wikipedia.org/wiki/Artificial_intelligence',\n    'https://en.wikipedia.org/wiki/Gradient_descent'\n]\nwith open('Wikipedia_scarp.txt','w') as f:\n  for url in urls:\n  # Send a GET request to the URL and retrieve the HTML content\n    response = requests.get(url)\n    html_content = response.content\n\n    # Parse the HTML content using Beautiful Soup\n    soup = BeautifulSoup(html_content, 'html.parser')\n\n    # Find the main content of the page (typically inside a div or section tag)\n    content_div = soup.find('div', {'id': 'mw-content-text'})\n\n    # Extract the text from the content div\n    text = content_div.get_text()\n    \n    # Print the scraped text\n    f.write(text)\n    \n\nprint('Scraped text has been written to file')\n"

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# This is the method where we manuallly write the names in the list
# Very much Boring.....


In [ ]:
'''
import requests
from bs4 import BeautifulSoup
import pandas as pd

base_url = "https://en.wikipedia.org/wiki/"

articles = ["Mahatma Gandhi","Data_science", "Python_(programming_language)","Machine_learning","Artificiall_intelligence"]

df = pd.DataFrame(columns=["Title", "Paragraphs"])

for article in articles[0:1]:
  url = base_url + article
  response = requests.get(url)
  html = response.content

  soup = BeautifulSoup(html,'html.parser')

  title = soup.find('h1',{'class':'firstHeading'}).text

  summary_div = soup.find('div', {'class': 'toc'})
  if summary_div is not None:
    summary = summary_div.text
  else:
    summary = ''


  text = soup.find('div',{'class':'mw-parser-output'}).text
  text_div = soup.find('div', {'class': 'mw-parser-output'})
  if text_div is not None:
    text = text_div.text
  else:
    text = ''

  paragraphs = []
  text_div = soup.find('div', {'class': 'mw-parser-output'})
  if text_div is not None:
    for p in text_div.find_all('p'):
      paragraphs.append(p.text)


  df.loc[len(df)] = [title, "\n".join(paragraphs)]

df.to_csv("wiki.csv",index = False)
'''

In [ ]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

base_url = "https://en.wikipedia.org/wiki/"

articles = ["Mahatma Gandhi","Data_science", "Python_(programming_language)","Machine_learning","Artificiall_intelligence"]

df = pd.DataFrame(columns=["Title", "Paragraphs"])

for article in articles:
  url = base_url + article
  response = requests.get(url)
  html = response.content

  soup = BeautifulSoup(html,'html.parser')

  title = soup.find('h1',{'class':'firstHeading'}).text

  paragraphs = []
  text_div = soup.find('div', {'class': 'mw-content-container'})
  if text_div is not None:
    for p in text_div.find_all('p'):
      paragraphs.append(p.text)

  df.loc[len(df)] = [title, "\n".join(paragraphs)]

df.to_csv("wiki.csv",index = False)


In [ ]:
from google.colab import files

df.to_csv("wikipedia_data.csv", index=False)

files.download("wikipedia_data.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
from google.colab import drive

drive.mount('/content/drive')

csv_path = "/content/drive/MyDrive/wikipedia_data.csv"

df.to_csv(csv_path, index=False)


Mounted at /content/drive


In [ ]:
import requests
import json

# Parameters for the API request
category = input("Enter the Wikipedia category: ")
url = "https://en.wikipedia.org/w/api.php"
params = {
    "action": "query",
    "format": "json",
    "list": "categorymembers",
    "cmtitle": category,
    "cmlimit": "max"
}

# Send API request and parse response
response = requests.get(url, params=params)
data = json.loads(response.text)

# Check if the response has a "query" key
if "query" in data:
    # Extract article titles from response
    articles = [article["title"] for article in data["query"]["categorymembers"]]

    # Save articles to file
    with open("articles.txt", "w") as f:
        for article in articles:
            f.write(article + "\n")

    print("List of articles saved to articles.txt")
else:
    print("No articles found in the specified category.")


In [ ]:
import os

# Save articles to file
file_path = os.path.join(os.getcwd(), "articles.txt")
with open("/content/drive/MyDrive/wikipedia_data.csv", "a") as f:
    for article in articles:
        f.write(article + "\n")


# But this method is pretty much good....

In [ ]:
from google.colab import drive
from bs4 import BeautifulSoup
import pandas as pd
import requests

drive.mount('/content/drive')

with open('/content/drive/MyDrive/articles.txt', 'r') as f:
    articles = [line.strip() for line in f.readlines()]

base_url = "https://en.wikipedia.org/wiki/"

df = pd.DataFrame(columns=["Title", "Paragraphs"])

for article in articles:
  url = base_url + article
  response = requests.get(url)
  html = response.content

  soup = BeautifulSoup(html,'html.parser')

  title = soup.find('h1',{'class':'firstHeading'}).text


  paragraphs = []
  text_div = soup.find('div', {'class': 'mw-content-container'})
  if text_div is not None:
    for p in text_div.find_all('p'):
      paragraphs.append(p.text)

  df.loc[len(df)] = [title, "\n".join(paragraphs)]

  if response.status_code != 200:
    print(f"The article '{article}' does not exists on Wikipedia.")
    articles.remove(article)
  #else:
    #print(f"The article '{article}' does not exist on Wikipedia.")

with open('/content/drive/MyDrive/articles.txt', 'w') as f:
    f.write('\n'.join(articles))

df.to_csv('/content/drive/MyDrive/wikipedia1.csv', index=False)


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# Working on this....to make more efficient🙃🙃🙃
All done...But don't use this code as it will not save the results directly in folder....
It is just like i used for refernce...

In [ ]:
import requests
import json
from bs4 import BeautifulSoup
import pandas as pd
from google.colab import drive

drive.mount('/content/drive')

category = input("Enter the Wikipedia category: ")
url = "https://en.wikipedia.org/w/api.php"
params = {
    "action": "query",
    "format": "json",
    "list": "categorymembers",
    "cmtitle": category,
    "cmlimit": "max"
}

response = requests.get(url, params=params)
data = json.loads(response.text)

if "query" in data:
    articles = [article["title"] for article in data["query"]["categorymembers"]]

    with open("articles.txt", "w") as f:
        for article in articles:
            f.write(article + "\n")

    print("List of articles saved to articles.txt")
else:
    print("No articles found in the specified category.")

with open('/content/drive/MyDrive/Wiki-datasets/articles.txt', 'r') as f:
    articles = [line.strip() for line in f.readlines()]

base_url = "https://en.wikipedia.org/wiki/"

df = pd.DataFrame(columns=[ "Title", "Paragraphs"])

for article in articles:
  url = base_url + article
  response = requests.get(url)
  html = response.content

  soup = BeautifulSoup(html,'html.parser')

  title = soup.find('h1',{'class':'firstHeading'}).text

  paragraphs = []
  text_div = soup.find('div', {'class': 'mw-content-container'})
  if text_div is not None:
    for p in text_div.find_all('p'):
      paragraphs.append(p.text)

  df.loc[len(df)] = [title, "\n".join(paragraphs)]

  if response.status_code != 200:
    print(f"The article '{article}' does not exist on Wikipedia.")
    df = df[df["Article"] != article]

df.to_csv('/content/drive/MyDrive/Wiki-datasets/wikipedia_data.csv', index=False)


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Enter the Wikipedia category: Category:Chemistry
List of articles saved to articles.txt


In [ ]:
!cp /content/drive/MyDrive/articles.txt /content/drive/MyDrive/Wiki-datasets

In [ ]:
!cp /content/drive/MyDrive/wikipedia_data.csv /content/drive/MyDrive/Wiki-datasets

In [ ]:
!cp /content/autoarticles.txt /content/drive/MyDrive

In [ ]:
!cp /content/autoarticles.txt /content/drive/MyDrive/autoarticles.txt

In [ ]:
!cp /content/drive/MyDrive/ColabNotebooks/WikipediaScraping.ipynb /content/drive/MyDrive/Wiki-datasets/WikipediaScraping.ipynb

# THIS IS WHAT U NEED.....Finallly. ✅
To take input use (Category:"SUBJECT U WANNA TAKE AS INPUT") and hit enter.
#EXPLAIN
in the below code it will directly store the name of articles of a particular subject without scraping the previous articles we had already scrapped....so the less chance to make Wikipedia 's server loaded with many requests...
#Limitations
it will not scrape the subject which have more than 5000 articles .i.e 5000 pages.
Aditionally, it's important to be mindful of Wikipedia's API usage limits and not to make too many requests too quickly, as this can result in your IP address being blocked.
#How to deal???
 example
####👉🏻United States: As of April 2023, the English-language Wikipedia has over 6.4 million articles, with the United States being one of the most extensively covered subjects.
####👉🏻Science: The field of science covers a broad range of subjects, including physics, chemistry, biology, astronomy, and more. Each of these subjects has a large number of articles and pages dedicated to them.
####👉🏻History: The subject of history covers a vast array of topics, with a large number of articles and pages dedicated to specific events, periods, and people throughout history.

####So now the question how to scrape it subject called Sciene here. Here's one way is to make a particualr subject an umbrella term for all the other subjects falling into that category. So here take astronomy as one subject and start scraping it and u will get desired results...🙃

####Another approach is to use a more powerful server or distributed computing to handle the scraping.But this approch i feel in itself is a limitation so....

In [ ]:
from google.colab import drive

drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import requests
import json
from bs4 import BeautifulSoup
import pandas as pd
from google.colab import drive
import os

drive.mount('/content/drive')

path = '/content/drive/MyDrive/Wiki-datasets/'

if os.path.isfile(path + 'society.txt'):

  with open(path + 'society.txt', 'r') as f:
    articles = [line.strip() for line in f.readlines()]
else:

  articles = []

category = input("Enter the Wikipedia category: ")
url = "https://en.wikipedia.org/w/api.php"
params = {
    "action": "query",
    "format": "json",
    "list": "categorymembers",
    "cmtitle": category,
    "cmlimit": "max"
}

response = requests.get(url, params=params)
data = json.loads(response.text)

if "query" in data:

    new_articles = [article["title"] for article in data["query"]["categorymembers"]]

    new_articles = list(set(new_articles) - set(articles))

    with open(path + "society.txt", "a") as f:
        for article in new_articles:
            f.write(article + "\n")

    print(f"{len(new_articles)} new articles found and saved to society.txt")
else:
    new_articles = []
    print("No articles found in the specified category.")

base_url = "https://en.wikipedia.org/wiki/"

df = pd.DataFrame(columns=[ "Title", "Paragraphs"])

for article in new_articles:
  url = base_url + article
  response = requests.get(url)
  html = response.content

  soup = BeautifulSoup(html,'html.parser')

  title = soup.find('h1',{'class':'firstHeading'}).text

  paragraphs = []
  text_div = soup.find('div', {'class': 'mw-content-container'})
  if text_div is not None:
    for p in text_div.find_all('p'):
      paragraphs.append(p.text)

  df.loc[len(df)] = [title, "\n".join(paragraphs)]

  if response.status_code != 200:
    print(f"The article '{article}' does not exist on Wikipedia.")
    df = df[df["Title"] != title]

df.to_csv('/content/drive/MyDrive/Wiki-datasets/society.csv', mode='a', header=not os.path.isfile('/content/drive/MyDrive/Wiki-datasets/society.csv'), index=False)


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Enter the Wikipedia category: Category:Stage theories
60 new articles found and saved to society.txt


###2nd method...Recursion
#####Another approach is to use a more powerful server or distributed computing to handle the scraping.But this approch i feel in itself is a limitation so....

In [ ]:
import requests
import json
from bs4 import BeautifulSoup
import pandas as pd
from google.colab import drive
import os

drive.mount('/content/drive')

path = '/content/drive/MyDrive/Wiki-datasets/'

if os.path.isfile(path + 'human_medicine.txt'):
    with open(path + 'human_medicine.txt', 'r') as f:
        articles = [line.strip() for line in f.readlines()]
else:
    articles = []

category = input("Enter the Wikipedia category: ")
base_url = "https://en.wikipedia.org/wiki/"

def scrape_articles_in_category(category):
    url = "https://en.wikipedia.org/w/api.php"
    params = {
        "action": "query",
        "format": "json",
        "list": "categorymembers",
        "cmtitle": category,
        "cmlimit": "max"
    }
    response = requests.get(url, params=params)
    data = json.loads(response.text)

    if "query" in data:
        new_articles = [article["title"] for article in data["query"]["categorymembers"]]
        new_articles = list(set(new_articles) - set(articles))
        with open(path + "human_medicine.txt", "a") as f:
            for article in new_articles:
                f.write(article + "\n")
        print(f"{len(new_articles)} new articles found and saved to human_medicine.txt")

        df = pd.DataFrame(columns=[ "Title", "Paragraphs"])
        for article in new_articles:
            url = base_url + article
            response = requests.get(url)
            html = response.content
            soup = BeautifulSoup(html,'html.parser')
            title = soup.find('h1',{'class':'firstHeading'}).text
            paragraphs = []
            text_div = soup.find('div', {'class': 'mw-content-container'})
            if text_div is not None:
                for p in text_div.find_all('p'):
                    paragraphs.append(p.text)
            df.loc[len(df)] = [title, "\n".join(paragraphs)]
            if response.status_code != 200:
                print(f"The article '{article}' does not exist on Wikipedia.")
                df = df[df["Title"] != title]                             # some error from this line need to be solved.....
        df.to_csv('/content/drive/MyDrive/Wiki-datasets/human_medicine.csv', mode='a', header=not os.path.isfile('/content/drive/MyDrive/Wiki-datasets/human_medicine.csv'), index=False)

    else:
        print("No articles found in the specified category.")

    subcategories = []
    for article in data["query"]["categorymembers"]:
        if article["ns"] == 14:
            subcategories.append(article["title"])

    for subcategory in subcategories:
        scrape_articles_in_category(subcategory)

scrape_articles_in_category(category)


  

###https://en.wikipedia.org/wiki/Wikipedia:Contents
#####use this link to directly scrape

In [ ]:
import pandas as pd
from google.colab import drive

drive.mount('/content/drive')
path = '/content/drive/MyDrive/Wiki-datasets/wikipedia_data.csv'

df = pd.read_csv(path)

print(df.head(10))

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
                             Title  \
0                   Mahatma Gandhi   
1  Assassination of Mahatma Gandhi   
2                   Gandhi Jayanti   
3      Criticism of Mahatma Gandhi   
4         Non-cooperation movement   
5             Champaran Satyagraha   
6                       Salt March   
7              Quit India Movement   
8                 Sabarmati Ashram   
9                       Satyagraha   

                                          Paragraphs  
0  \n\nMohandas Karamchand Gandhi (/ˈɡɑːndi, ˈɡæn...  
1  \n\nArrested\n\nMahatma Gandhi was assassinate...  
2  \n\nGandhi Jayanti is an event celebrated in I...  
3     Other reasons this message may be displayed:\n  
4  \n\nThe Non-cooperation movement was a politic...  
5  \n\nThe Champaran Satyagraha of 1917 was the f...  
6  \n\nThe Salt March, also known as the Salt Sat...  
7  \n\nMahat

In [ ]:
import pandas as pd

df = pd.read_csv('/content/drive/MyDrive/Wiki-datasets/wikipedia_data.csv')
article_title = 'Ira V. Hiscock'
article_df = df.loc[df['Title'] == article_title]

print(article_df['Title'].values[0])
print(article_df['Paragraphs'].values[0])

In [ ]:
#----> But still see how we will deal with above mathematical ariticle. How we will deal with this?? If we delete this type of data some data will be lost....

#Paragraph-->Sentence

some inspiration:
Tokenization for Natural Language Processing

Towards Data Science
https://towardsdatascience.com/tokenization-for-natural-language-processing-a179a891bad4

nltk documentation https://www.nltk.org/

In [ ]:
import nltk
nltk.download('punkt')
from nltk import sent_tokenize

text = "Thiis is an example paragraph. It contains multiple sentences. The sentences will be split into a list. "
sentences = sent_tokenize(text)
print(sentences)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


['Thiis is an example paragraph.', 'It contains multiple sentences.', 'The sentences will be split into a list.']


In [ ]:
import pandas as pd
import nltk

nltk.download('punkt')

df = pd.read_csv('/content/drive/MyDrive/wikipedia1.csv')

df['Sentences'] = df['Paragraphs'].apply(nltk.sent_tokenize)

df = df.explode('Sentences')

df = df.drop(columns=['Paragraphs'])

df.to_csv('/content/drive/MyDrive/wikipedia1.csv', index=False)


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [ ]:
print(df.columns)

Index(['Title', 'Sentences'], dtype='object')


In [ ]:
!cp /content/drive/MyDrive/Wiki-datasets/culture.csv /content/drive/MyDrive/Wiki-datasets/culture_copy.csv

In [ ]:
!cp /content/drive/MyDrive/Wiki-datasets/geography.csv /content/drive/MyDrive/Wiki-datasets/geography_copy.csv

In [ ]:
!cp /content/drive/MyDrive/Wiki-datasets/health.csv /content/drive/MyDrive/Wiki-datasets/health_copy.csv

In [ ]:
!cp /content/drive/MyDrive/Wiki-datasets/nature.csv /content/drive/MyDrive/Wiki-datasets/nature_copy.csv

In [ ]:
!cp /content/drive/MyDrive/Wiki-datasets/people.csv /content/drive/MyDrive/Wiki-datasets/people_copy.csv

In [ ]:
!cp /content/drive/MyDrive/Wiki-datasets/philosophy.csv /content/drive/MyDrive/Wiki-datasets/philosophy_copy.csv

In [ ]:
!cp /content/drive/MyDrive/Wiki-datasets/religion.csv /content/drive/MyDrive/Wiki-datasets/religion_copy.csv

In [ ]:
!cp /content/drive/MyDrive/Wiki-datasets/society.csv /content/drive/MyDrive/Wiki-datasets/society_copy.csv

In [ ]:
import pandas as pd
import nltk
import re

nltk.download('punkt')

df = pd.read_csv('/content/drive/MyDrive/Wiki-datasets/society_copy.csv')

df['Sentences'] = df['Paragraphs'].apply(lambda x: nltk.sent_tokenize(str(x)))

pattern = r'\[\d+\]'

df['Sentences'] = df['Sentences'].apply(lambda sentence_list: [re.sub(pattern, '', sentence).strip() for sentence in sentence_list])

df = df.explode('Sentences')

df = df.drop(columns=['Paragraphs'])

df.to_csv('/content/drive/MyDrive/Wiki-datasets/society_copy.csv', index=False)


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [ ]:
import pandas as pd

csv_files = ['/content/drive/MyDrive/Wiki-datasets/culture_copy.csv', '/content/drive/MyDrive/Wiki-datasets/geography_copy.csv', '/content/drive/MyDrive/Wiki-datasets/health_copy.csv','/content/drive/MyDrive/Wiki-datasets/nature_copy.csv','/content/drive/MyDrive/Wiki-datasets/people_copy.csv','/content/drive/MyDrive/Wiki-datasets/philosophy_copy.csv','/content/drive/MyDrive/Wiki-datasets/religion_copy.csv','/content/drive/MyDrive/Wiki-datasets/society_copy.csv']

merged_data = pd.DataFrame()

for file in csv_files:
    data = pd.read_csv(file)

    data['File'] = file

    merged_data = pd.concat([merged_data, data], ignore_index=True)

merged_data.to_csv('/content/drive/MyDrive/Wiki-datasets/merged_data.csv', index=False)


In [ ]:
import pandas as pd

file_path = '/content/drive/MyDrive/Wiki-datasets/merged_data.csv'
data = pd.read_csv(file_path)
data.head(10)

In [ ]:
import pandas as pd
import os

csv_files = ['/content/drive/MyDrive/Wiki-datasets/culture_copy.csv', '/content/drive/MyDrive/Wiki-datasets/geography_copy.csv', '/content/drive/MyDrive/Wiki-datasets/health_copy.csv','/content/drive/MyDrive/Wiki-datasets/nature_copy.csv','/content/drive/MyDrive/Wiki-datasets/people_copy.csv','/content/drive/MyDrive/Wiki-datasets/philosophy_copy.csv','/content/drive/MyDrive/Wiki-datasets/religion_copy.csv','/content/drive/MyDrive/Wiki-datasets/society_copy.csv']

merged_data = pd.DataFrame()

for file in csv_files:
    data = pd.read_csv(file)

    file_name = os.path.splitext(os.path.basename(file))[0]
    file_type = file_name.split('_')[0]
    data['File'] = file_type

    merged_data = pd.concat([merged_data, data], ignore_index=True)

merged_data.to_csv('/content/drive/MyDrive/Wiki-datasets/merged_data1.csv', index=False)


In [ ]:
import pandas as pd

file_path = '/content/drive/MyDrive/Wiki-datasets/merged_data1.csv'
data = pd.read_csv(file_path)
data.tail(10)

In [ ]:
middle_data = data.head(4).append(data.tail(7))

In [ ]:
middle_data

In [ ]:
head_rows = data.head(5).values.tolist()

head_rows

In [ ]:
tail_rows = data.tail(10).values.tolist()

tail_rows

# Data Preprocessing

In [ ]:
import pandas as pd

data = pd.read_csv('/content/drive/MyDrive/Wiki-datasets/merged_data1.csv')
data.head()

Title                                          Sentences     File
0  V. Kalyanam  V. Kalyanam (15 August 1922 – 4 May 2021) was ...  culture
1  V. Kalyanam  He joined the freedom struggle during Quit Ind...  culture
2  V. Kalyanam  Kalyanam was just behind Gandhi when Nathuram ...  culture
3  V. Kalyanam  According to Kalyanam, Gandhi died instantly a...  culture
4  V. Kalyanam  He was the first to inform Nehru and Patel abo...  culture

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3373599 entries, 0 to 3373598
Data columns (total 3 columns):
 #   Column     Dtype 
---  ------     ----- 
 0   Title      object
 1   Sentences  object
 2   File       object
dtypes: object(3)
memory usage: 77.2+ MB


In [ ]:
data.describe()

Title  \
count                                             3373599   
unique                                              66986   
top     List of Supernatural and The Winchesters chara...   
freq                                                 2900   

                                      Sentences     File  
count                                   3348337  3373599  
unique                                  3007849        8  
top     You can help Wikipedia by expanding it.   health  
freq                                      11412   516265

In [ ]:
import re
import nltk
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from nltk.tokenize import word_tokenize
import pickle

class Preprocessing:
  def __init__(self, num_words, seq_len):
    #self.data = '/content/drive/MyDrive/Wiki-datasets/merged_data1.csv'
    self.data = '/content/drive/MyDrive/Turing_science_projects_23/Neural language models/Wiki-datasets/merged_data1.csv'
    self.num_words = num_words
    self.seq_len = seq_len
    self.vocabulary = None
    self.x_tokenized = None
    self.x_padded = None
    self.x_raw = None
    self.y = None

    self.x_train = None
    self.x_test = None
    self.y_train = None
    self.y_test = None

  def load_data(self):
    # Reads the raw csv file and split into
    # sentences (x) and target (y)
    df = pd.read_csv(self.data)
    #df.drop(['id','keyword','location'], axis=1, inplace=True)
    predictors = ['Sentences','Title']
    self.x_raw = df[predictors].values
    self.y = df['File'].values
  '''
  def clean_text(self):
    # Removes special symbols and just keep
    # words in lower or upper form
    self.x_raw = [x.lower() for x in self.x_raw]
    self.x_raw = [re.sub(r'[^A-Za-z]+', ' ', x) for x in self.x_raw]
  '''
  def clean_text(self):
    # Removes special symbols and just keep
    # words in lower or upper form
    self.x_raw = [str(x).lower() for x in self.x_raw]
    self.x_raw = [re.sub(r'[^A-Za-z]+', ' ', x) for x in self.x_raw]


  def text_tokenization(self):
    # Tokenizes each sentence by implementing the nltk tool
    self.x_raw = [word_tokenize(x) for x in self.x_raw]

  def build_vocabulary(self):
    # Builds the vocabulary and keeps the "x" most frequent word
    self.vocabulary = dict()
    fdist = nltk.FreqDist()

    for sentence in self.x_raw:
      for word in sentence:
        fdist[word] += 1

    common_words = fdist.most_common(self.num_words)

    for idx, word in enumerate(common_words):
      self.vocabulary[word[0]] = (idx+1)

  def word_to_idx(self):
    # By using the dictionary (vocabulary), it is transformed
    # each token into its index based representatio
    self.x_tokenized = list()

    for sentence in self.x_raw:
      temp_sentence = list()
      for word in sentence:
        if word in self.vocabulary.keys():
          temp_sentence.append(self.vocabulary[word])
      self.x_tokenized.append(temp_sentence)

  def padding_sentences(self):
    # Each sentence which does not fulfill the required le
    # it's padded with the index 0
    pad_idx = 0
    self.x_padded = list()

    for sentence in self.x_tokenized:
      while len(sentence) < self.seq_len:
        sentence.insert(len(sentence), pad_idx)
      self.x_padded.append(sentence)

    self.x_padded = np.array(self.x_padded)

  def split_data(self):
    self.x_train, self.x_test, self.y_train, self.y_test = train_test_split(self.x_padded, self.y, test_size=0.2, random_state=42)

  def save_data(self, filename):
    data = {
        'x_train': self.x_train,
        'x_test': self.x_test,
        'y_train': self.y_train,
        'y_test': self.y_test,
        'vocabulary': self.vocabulary
    }
    with open(filename, 'wb') as file:
        pickle.dump(data, file)



In [ ]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [ ]:
preprocessor = Preprocessing(num_words=500000, seq_len=50)
preprocessor.load_data()
preprocessor.clean_text()
preprocessor.text_tokenization()
preprocessor.build_vocabulary()
preprocessor.word_to_idx()
preprocessor.padding_sentences()
preprocessor.split_data()
#preprocessor.save_data('/content/drive/MyDrive/Wiki-datasets/preprocessing_data.pkl')
preprocessor.save_data('/content/drive/MyDrive/Turing_science_projects_23/Neural language models/Wiki-datasets/preprocessing_data5.pkl')
''' num_words = 4000: preprocessing_data2.pkl
num_words = 50000: preprocessing_data3.pkl
num_words = 250000: preprocessing_data4.pkl
num_words = 500000: preprocessing_data5.pkl '''

<ipython-input-2-750ee3e073ec>:89: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  self.x_padded = np.array(self.x_padded)


' num_words = 4000: preprocessing_data2.pkl\nnum_words = 50000: preprocessing_data3.pkl\nnum_words = 250000: preprocessing_data4.pkl\nnum_words = 500000: preprocessing_data5.pkl '

In [ ]:
import pickle

# Path to the pickle file
#file_path = '/content/drive/MyDrive/Wiki-datasets/preprocessing_data.pkl'
file_path = '/content/drive/MyDrive/Turing_science_projects_23/Neural language models/Wiki-datasets/preprocessing_data5.pkl'

# Load the data from the pickle file
with open(file_path, 'rb') as file:
    data = pickle.load(file)

# Print the available keys in the data dictionary
print(data.keys())


dict_keys(['x_train', 'x_test', 'y_train', 'y_test', 'vocabulary'])


In [ ]:
import pickle

# Path to the pickle file
#file_path = '/content/drive/MyDrive/Wiki-datasets/preprocessing_data.pkl'
file_path = '/content/drive/MyDrive/Turing_science_projects_23/Neural language models/Wiki-datasets/preprocessing_data5.pkl'

# Load the data from the pickle file
with open(file_path, 'rb') as file:
    data = pickle.load(file)

# Access a subset of the data

subset_data = data['x_test'][:20]

# Print the subset of the data
for item in subset_data:
    print(item)


[1, 271, 3, 232, 96, 523, 5, 4284, 6, 442, 2, 150, 4581, 3, 15733, 206, 1628, 123, 4562, 4, 2478, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[2489, 326, 14, 1, 60, 207, 346, 2, 82, 2189, 332, 326, 114, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[15, 857, 6617, 1987, 2, 1, 1824, 492, 2542, 73, 301, 9, 12603, 2, 1, 3473, 11774, 822, 45, 4, 557, 7, 13602, 4, 18, 11788, 11, 1, 1824, 12603, 9927, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[36, 905, 7, 9, 6, 1048, 2, 4521, 1346, 1713, 3951, 3, 19601, 34, 278, 2031, 38, 33, 2481, 45, 5324, 853, 33, 692, 374, 153, 5, 6746, 2275, 14726, 8, 2234, 576, 2052, 374, 3650, 3, 1, 20343, 7359, 886, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[1, 6124, 2532, 5282, 77, 19, 1337, 21252, 5719, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[1, 214990, 74, 6726, 1, 15789, 8, 

In [ ]:
import pandas as pd

# Generator function to read the file in chunks
def read_data(filename, chunksize):
    for chunk in pd.read_csv(filename, chunksize=chunksize):
        yield chunk

# Define the file path and chunk size
filename = "/content/drive/MyDrive/Wiki-datasets/merged_data1.csv"
chunksize = 100000

# Initialize an empty set to store unique words
unique_words = set()

# Iterate over chunks of the data
for data_chunk in read_data(filename, chunksize):
    sentences = data_chunk["Sentences"]
    words = " ".join(str(sentence) for sentence in sentences).split()
    unique_words = set(words)


# Count the number of unique words
num_unique_words = len(unique_words)

print("Number of different words in the corpus:", num_unique_words)


Number of different words in the corpus: 133248


Number of different words in the corpus: 57919   chunk= 50000

---
Number of different words in the corpus: 93876   chunk= 90000

---
Number of different words in the corpus: 133248   chunk = 100000




In [ ]:
import pandas as pd

# Load the data into a DataFrame
data = pd.read_csv("/content/drive/MyDrive/Wiki-datasets/merged_data1.csv")

# Access the "Sentences" column
sentences = data["Sentences"].astype(str)

# Split each sentence into words
all_words = " ".join(sentences).split()

# Count the number of unique words
num_unique_words = len(set(all_words))

print("Number of different words in the corpus:", num_unique_words)


Number of different words in the corpus: 2058601
